File: leetcode-javascript/go/54.go
package main

import "fmt"

func spiralOrder(matrix [][]int) []int {
	m := len(matrix)
	n := len(matrix[0])

	res := make([]int, m*n)
	var in int
	// u 上边界 l 左边界， d 下边界 r 右边界
	u, l := 0, 0
	d, r := m-1, n-1
	for {
		for i := l; i <= r; i++ { //  上边界遍历
			res[in] = matrix[u][i]
			in++
		}
		u++
		if u > d {
			break
		}

		for i := u; i <= d; i++ { //  右边界遍历
			res[in] = matrix[i][r]
			in++
		}
		r--
		if l > r {
			break
		}

		for i := r; i >= l; i-- { //  下边界遍历
			res[in] = matrix[d][i]
			in++
		}
		d--
		if u > d {
			break
		}

		for i := d; i >= u; i-- { //  右边界遍历
			res[in] = matrix[i][l]
			in++
		}
		l++
		if l > r {
			break
		}

	}

	return res
}

func main() {
	values := [][]int{}

	// Step 2: 使用 appped() 函数向空的二维数组添加两行一维数组
	row1 := []int{2, 5, 8}
	row2 := []int{4, 0, -1}
	values = append(values, row1)
	values = append(values, row2)
	fmt.Printf("#-->%v\n", spiralOrder(values))
}

File: leetcode-javascript/54.js
/**
 * 给你一个 m 行 n 列的矩阵 matrix ，请按照 顺时针螺旋顺序 ，返回矩阵中的所有元素。

 

示例 1：


输入：matrix = [[1,2,3],[4,5,6],[7,8,9]]
输出：[1,2,3,6,9,8,7,4,5]
示例 2：


输入：matrix = [[1,2,3,4],[5,6,7,8],[9,10,11,12]]
输出：[1,2,3,4,8,12,11,10,9,5,6,7]

来源：力扣（LeetCode）
链接：https://leetcode-cn.com/problems/spiral-matrix
著作权归领扣网络所有。商业转载请联系官方授权，非商业转载请注明出处。
 */
/**
 * @param {number[][]} matrix
 * @return {number[]}
 */
 var spiralOrder = function(matrix) {
   if (matrix.length === 0) return []
   let t = 0
   let b = matrix.length - 1
   let l = 0
   let r = matrix[0].length - 1
   let res = []
   while(true) {
     for (let i = l; i <= r; ++i) res.push(matrix[t][i])
     if (++t > b) break
     for (let i = t; i <= b; ++i) res.push(matrix[i][r])
     if (--r < l) break
     for (let i = r; i >= l; i--) res.push(matrix[b][i])
     if (--b < t) break
     for (let i = b; i >= t; --i) res.push(matrix[i][l])
     if (++l > r) break
   }
   return res
};

// console.log(spiralOrder([[1,2,3],[4,5,6],[7,8,9]]))
// console.log(spiralOrder([[1,2,3,4],[5,6,7,8],[9,10,11,12]]))
// console.log(spiralOrder([[2,5],[8,4],[0,-1]]))
// console.log(spiralOrder([[1,2,3],[4,5,6],[7,8,9]]))
console.log(spiralOrder([[2,5,8],[4,0,-1]]))
// console.log(spiralOrder([[]]))